pkg/webhook/pod/pd_deleter.go

```diff
@@ -0,0 +1,175 @@
+		return pc.admitDeletePDLeader(pod, tc, pdClient, ordinal)
+	}
+
+	klog.Infof("pod[%s/%s] is not pd-leader,admit to delete", namespace, name)
+	return util.ARSuccess()
+}
+
+// this pod is not a member of pd cluster currently, it could be deleted from pd cluster already or haven't registered in pd cluster
+// we need to check whether this pd pod has been ensured wouldn't be a member in pd cluster
+func (pc *PodAdmissionControl) admitDeleteNonPDMemberPod(IsDeferDeleting, isInOrdinal bool, pod *corev1.Pod, ownerStatefulSet *apps.StatefulSet, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient, ordinal int32) *v1beta1.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+	tcName := tc.Name
+
+	// check whether this pod has been ensured wouldn't be a member in pd cluster
+	if IsDeferDeleting {
+		// when pd scale in, we should delete member first and finally edit its pvc and admit to delete pod.
+		// If we edit pvc first and finally delete pd member during scale in,
+		// it can be error that pd scale in From 4 to 3,the pvc was edited successfully and
+		// the pd member was fail to deleted. If the pd scale was recover to 4 at that time,
+		// it would be existed an pd-3 instance with its deferDeleting label Annotations PVC.
+		// And the pvc can be deleted during upgrading if we use create pod webhook in future.
+		if !isInOrdinal {
+			err := addDeferDeletingToPVC(pc, tc, ownerStatefulSet.Name, namespace, ordinal)
+			if err != nil {
+				klog.Infof("tc[%s/%s]'s pod[%s/%s] failed to update pvc,%v", namespace, tcName, namespace, name, err)
+				return util.ARFail(err)
+			}
+		}
+		klog.Infof("pd pod[%s/%s] is not member of tc[%s/%s],admit to delete", namespace, name, namespace, tcName)
+		return util.ARSuccess()
+
+	}
+	err := pdClient.DeleteMember(name)
+	if err != nil {
+		return util.ARFail(err)
+	}
+	err = addDeferDeletingToPDPod(pc, pod)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	// make sure this pd pod won't be a member of pd cluster any more
+	return &v1beta1.AdmissionResponse{
+		Allowed: false,
+	}
+}
+
+func (pc *PodAdmissionControl) admitDeleteExceedReplicasPDPod(pod *corev1.Pod, tcName string, pdClient pdapi.PDClient) *v1beta1.AdmissionResponse {
+
+	name := pod.Name
+	namespace := pod.Namespace
+
+	err := pdClient.DeleteMember(name)
+	if err != nil {
+		return util.ARFail(err)
+	}
+	// we should add deferDeleting Annotation when we delete member successfully.
+	err = addDeferDeletingToPDPod(pc, pod)
+	if err != nil {
+		return util.ARFail(err)
+	}
+
+	klog.Infof("tc[%s/%s]'s pd[%s/%s] is being deleted from pd-cluster,refuse to delete it.", namespace, tcName, namespace, name)
+	return &v1beta1.AdmissionResponse{
+		Allowed: false,
+	}
+}
+
+// this pod is a pd leader, we should transfer pd leader to other pd pod before it gets deleted before.
+func (pc *PodAdmissionControl) admitDeletePDLeader(pod *corev1.Pod, tc *v1alpha1.TidbCluster, pdClient pdapi.PDClient, ordinal int32) *v1beta1.AdmissionResponse {
```

Comment:
TransferPDLeader

Reply:
fix in https://github.com/pingcap/tidb-operator/pull/1114